cadence/contracts/AxelarGateway.cdc

```diff
@@ -0,0 +1,312 @@
+    ) {
+      self.sourceChain = sourceChain
+      self.sourceAddress = sourceAddress
+      self.contractAddress = contractAddress
+      self.payloadHash = payloadHash
+      self.sourceTxHash = sourceTxHash
+      self.sourceEventIndex = sourceEventIndex
+    }
+  }
+
+  access(all) let executedCommandIds: [String]
+  access(self) let approvedCommands: @{String: CGPCommand}
+
+  init() {        
+    self.SELECTOR_TRANSFER_OPERATORSHIP = Crypto.hash("transferOperatorship".utf8, algorithm: HashAlgorithm.KECCAK_256)
+    self.SELECTOR_APPROVE_CONTRACT_CALL = Crypto.hash("approveContractCall".utf8, algorithm: HashAlgorithm.KECCAK_256)
+    self.PREFIX_APP_CAPABILITY_NAME = "AppCapabilityPath"
+
+    self.executedCommandIds = []
+    self.approvedCommands <- {}
+  }
+
+  access(all) struct ParamHandler {
+    access(all) let rawParams: [AnyStruct]
+
+    init(params: [AnyStruct]) {
+      self.rawParams = params
+    }
+
+    access(contract) fun generateApproveContractCallParams(): ApproveContractCallParams? {
+      let sourceChain = self.rawParams[0].isInstance(Type<String>())
+      let sourceAddress = self.rawParams[1].isInstance(Type<String>())
+      let contractAddress = self.rawParams[2].isInstance(Type<String>())
+      let payloadHash = self.rawParams[3].isInstance(Type<String>())
+      let sourceTxHash = self.rawParams[4].isInstance(Type<String>())
+      let sourceEventIndex = self.rawParams[5].isInstance(Type<UInt256>())
+
+      if !sourceChain || !sourceAddress || !contractAddress || !payloadHash || !sourceTxHash || !sourceEventIndex {
+        return nil
+      }
+
+      return ApproveContractCallParams(
+        sourceChain : self.rawParams[0] as! String,
+        sourceAddress : self.rawParams[1] as! String,
+        contractAddress : self.rawParams[2] as! String,
+        payloadHash : self.rawParams[3] as! String,
+        sourceTxHash : self.rawParams[4] as! String,
+        sourceEventIndex : self.rawParams[5] as! UInt256,
+      )
+    }
+
+    access(contract) fun generateTransferOperatorshipParams(): AxelarAuthWeighted.TransferOperatorshipParams? {
+      let newOperators = self.rawParams[0].isInstance(Type<[String]>())
+      let newWeights = self.rawParams[1].isInstance(Type<[UInt256]>())
+      let newThreshold = self.rawParams[2].isInstance(Type<UInt256>())
+
+      if !newOperators || !newWeights || !newThreshold {
+        return nil
+      }
+
+      return AxelarAuthWeighted.TransferOperatorshipParams(
+        newOperators : self.rawParams[0] as! [String],
+        newWeights : self.rawParams[1] as! [UInt256],
+        newThreshold : self.rawParams[2] as! UInt256,
+      )
+    }
+  }
+
+  access(all) resource interface Executable {
+    access(all) fun executeApp(commandResource: &AxelarGateway.CGPCommand, sourceChain: String, sourceAddress: String, payload: [UInt8])
+  }
+
+  access(all) fun callContract(sender: Address, destinationChain: String, destinationContractAddress: String, payload: [UInt8]) {
```

Comment:
Is `sender` an important value here? There can be no guarantees that the provided address is accurate as a caller can define anyone as sender. While it's generally an anti-pattern in Cadence, happy to discuss how we can bridge identity-based patterns found in the source Solidity docs with the capability-based security in Cadence. As mentioned in #3, the current approach to sender addresses will be very brittle and requires trusting the caller to be honest.

Reply:
Sender is important as it allows the dApp to verify that the call is coming from their own internal contracts or known contracts. We are thinking of taking in a reference to the `AxelarGateway.Sender` capability where we can extract the address from the capability. Would this be a feasible approach?